lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -428,91 +434,178 @@ static SILValue emitCodeForSymbolicValue(SymbolicValue symVal,
   }
 }
 
-/// Collect the end-of-lifetime instructions of the given SILValue. These are
-/// either release_value or destroy_value instructions.
-/// \param value SIL value whose end-of-lifetime instructions must be collected.
-/// \param lifetimeEndInsts buffer for storing the found end-of-lifetime
-/// instructions of 'value'.
-static void getLifetimeEndInstructionsOfSILValue(
-    SILValue value, SmallVectorImpl<SILInstruction *> &lifetimeEndInsts) {
-
-  bool continueLifetimeEndInstructionSearch = true;
-  SILValue currValue = value;
-
-  while (continueLifetimeEndInstructionSearch) {
-    continueLifetimeEndInstructionSearch = false;
-
+/// Collect the instructions that mark the end of use of the given
+/// non-trivial SILValue at plus one (or owned if in OSSA).
+/// \param value non-trivial SILValue at plus one.
+/// \param fun SILFunction containing the SILValue \c value
+/// \param endUsers buffer for storing the found end users of the SILValue.
+static void
+getEndUsersOfPlusOneSILValue(SILValue value, SILFunction *fun,
+                             SmallVectorImpl<SILInstruction *> &endUsers) {
+  assert(!value->getType().isTrivial(*fun));
+
+  // Step 1. Collect a list of potential end users of value using a worklist.
+  SmallPtrSet<SILInstruction *, 4> potentialEndUsers;
+  SmallVector<SILValue, 4> worklist;
+  worklist.push_back(value);
+
+  while (!worklist.empty()) {
+    SILValue currValue = worklist.pop_back_val();
+    bool isTrivial = currValue->getType().isTrivial(*fun);
     for (Operand *use : currValue->getUses()) {
       SILInstruction *user = use->getUser();
-
-      if (isa<ReleaseValueInst>(user) || isa<DestroyValueInst>(user)) {
-        lifetimeEndInsts.push_back(user);
+      // If the user creates a derived, dependent value, include it in the
+      // worklist.
+      if (isa<CopyValueInst>(user) || isa<DestructureStructInst>(user) ||
+          isa<DestructureTupleInst>(user)) {
+        for (SILValue result : user->getResults()) {
+          worklist.push_back(result);
+        }
+      }
+      // For trivial values any use can be a potential end use, for non-trivial
+      // values only consider consuming uses as potential end uses. (Note that
+      // trivial values may arise in this search due to destructure_struct
+      // instructions.)
+      if (isTrivial) {
```

Comment:
Does this deal with addresses? If you are just worried about trivial values from destructures, I would just not add them to the worklist when adding results. Your thoughts?

Reply:
No it doesn't deal with addresses (see the above explanation). If there is a utility that does this including dealing with addresses it would be great. But, at this point it would be a overkill for OSLog, it wouldn't hurt to use that utility especially if it brings a conceptual clarity.

Reply:
I wasn't expecting it to handle addresses. I was more asking if there was a reason why you were checking for trivial here. Seems like you dont need to visit the trivial things.

Reply:
As I mentioned above, I do need to track at dependencies of trivial values. This pass is interested in them. It is supposed to track end of data dependent chains.  I am updating this code to simplify and generalize it.